### ISSUE TITLE
Consider removing Generator::schedule() and ScheduleParam<>

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#2552]
These were meant to make it feasible to write Generators that could be reused as part of a complex pipeline, by allowing code to make decisions based on parent code that was not yet well-defined (e.g., compute and store levels).

However, the advent of deferred-value LoopLevel (https://github.com/halide/Halide/pull/2504) seems to have made these constructs unnecessary, and arguably more trouble than they are worth:

- Analysis of some extremely complex Halide pipelines inside Google showed that the `schedule()` method was totally unnecessary, as the only deferred-value parameters needed turned out the be deferred-value LoopLevels.
- Splitting Halide pipelines into `generate()` and `schedule()` sections has proven to be unpopular and hard to explain; if deferred-value LoopLevel has made this division not-strictly-necessary, the only argument for keeping it is one of code style rather than code functionality. (And of course, for code style, the author can always opt-in to structuring her code in this way if they like.)
- We've seen exactly zero real-world use cases for the deferred-scalar-value `ScheduleParam<Expr>` construct; every possible case we've seen can be satisfactorily handled by an ordinary `GeneratorParam`.
- Finally, this ongoing experiment has led to a lot of churn and confusion about the status of Generator, which is detrimental to using it more widely.

I'd like to propose that we simplify Generator as follows, with a goal of locking down Generator design for the forseeable future:
- Replace `ScheduleParam<LoopLevel>` with `GeneratorParam<LoopLevel>`; since all `LoopLevel` instances are now deferred-value, we don't need a separate construct to imply this.
- Eliminate `ScheduleParam` entirely.
- Eliminate `Generator::schedule()` entirely.
- Revise `Stub` usage to support a static-call-only mode, eliminating entirely the stateful-class-instance code we have had up to now; this basically transmutes the stub code into a simple, function-style wrapper that instantiates and calls a Generator, while still hiding the Generator implementation details from the caller without requiring refactoring by the coder.

If you write or consume Halide code in Generator form, we'd very much like your feedback on this. (Pre-emptively tagging @abadams @dsharletg @zvookin because I know they have particular interests in the matter.)

**Comment by zvookin** [comment]
I'd like to capture a bit of context for history here. There were a number of reasons for introducing `schedule()` in the first place:

- We present the separation of algorithm and schedule as a fundamental premise of Halide and state that many schedules can be used with a single algorithm. As such, separating them in the code seems to make sense.The reality around this turns out to be somewhat more complex however. First, there are cases where the algorithm and schedule are intimately tied together and the algorithm ends up changing at least a little bit to facilitate a particular executions strategy. This happens often enough that it needs to be handled cleanly.

    Second separating the two requires structuring things such that schedulable Funcs and Vars, etc. are available at class scope. I consider this an improvement to readability where others seem to consider it tedious.

    Third, there's likely a bit of a difference of opinion around whether scheduling is a main part of Halide programming, or a necessary evil that one must do in order to make the code run fast. My position is separating the two makes the algorithm easier to read and the schedule easier to evolve, if harder to write. But there is an argument to be made that seeing the schedule next to the code allows getting the whole thing in one's head more easily.

   Point being that even if there is not direct requirement to separate scheduling, it may be the best way to write code. There does not appear to be consensus on this which makes me question the very being of separation of algorithm and schedule.

- With the auto scheduler we need to get rid of the manual scheduling. Not applying it at all seems the best way to do this:

 ```
   something_inside_infrastructure() {
        auto f = generate();
        if (auto_scheduling) {
            auto_schedule(f);
        } else {
            f.schedule();
        }
    }
```
- Building the IR and scheduling it must be done in two passes because one can have `Func outer` calling `Func inner` but with `inner.compute_at(outer, ...)`. `outer` cannot exist until `inner` is built and thus the scheduling has to happen after everything is built. The idea was this would be explicitly exposed with`generate()` being the first pass and `schedule()` being the second. Deferred loop levels move the second pass into Halide lowering which eliminates the main instance of this issue. Other uses of the two passes didn't really materialize and we've decided the design is more antipattern than pattern.

It is true there was push back on putting scheduling in `schedule()`. I'll let those who don't like that approach weigh in if they want to. I'm not sure it is hard to explain in and unto itself. In fact I think it is a great improvement in writing Halide code in many ways. There were issues around having to call schedule when instantiating a generator, but I think all of that goes away if we eliminate schedule params and the exposed two pass approach.

I fully realize that I am swimming against the tide on this. However characterizing `schedule()` as "unnecessary" strikes me as a questionable framing. Comments are unnecessary too yet I doubt any of us would think highly of a language that didn't have them.

More concretely actionable stuff:

- I'm not sure `GeneratorParam<LoopLevel>` is the right thing as loop levels are generally not configured in the same way other `GeneratorParams` are. Perhaps giving it its own parameter type? E.g. We could try to bring the word "at" in to tie it to `compute_at` and `store_at` e.g. `AtLoopLevel`. One could prototype that via `typedef GeneratorParam<LoopLevel> AtLoopLevel` to see if it makes for better language.

- What is the plan for auto scheduling vs. manual scheduling?

- It seems simpler to eliminate the schedule params and the schedule method in two steps anyway. If there is no support for `schedule()` and there is an acceptable answer for auto scheduling, I'll let it go, but I really think we need more experience writing composable generators before we'll know the right answer here.

**Comment by steven-johnson** [comment]
> I'm not sure GeneratorParam<LoopLevel> is the right thing as 
> loop levels are generally not configured in the same way other GeneratorParams are.

I don't disagree, but I think it's less-weird than other options: it means that all compile-time parameters to a Generator are uniformly specified by `GeneratorParam`, period. Yes, `LoopLevel` would be a bit of a special case, but we already have precedent for special cases there (e.g. GeneratorParam<Target> target` being magically set by Halide for you).

> It seems simpler to eliminate the schedule params and the schedule method in two steps anyway.

Agree 100% -- if we went down this road it would definitely be a staged thing.

**Comment by abadams** [comment]
My issue with the separate schedule() method in practice was just that it required declaring every Func far from its definition, and every RDom far from its use, which I found tedious and ugly. I think separation of algorithm from schedule should be done by placing the schedule in a nested scope, or equivalently, after all the algorithm code, to reflect the fact that the schedule refers to the internals of the algorithm, but the algorithm may not refer to the schedule. I don't disagree with your other points, and this is clearly a matter of taste, so I don't want to make a strong case out of the readability angle.

The bigger issue in practice was legacy baggage. The separation made it impossible to upgrade to the new Generator and deprecate the old one because we would have had to untangle the mixture of algorithm and schedule in every current pipeline. The other changes to Generator have a much easier upgrade path.

Basically we tried separating them, and as a result adoption of new Generator features stalled entirely. 

For this and several other reasons, (the need to provide estimates, long compile-times, potentially larger-than-reasonable performance regressions on compiler updates, painting the autoscheduler into a corner where we can't change it), I don't think we'll be able to turn on the autoscheduler for pipelines not expecting it, which I think is the non-readability reason for a separate schedule() method. I think pipelines will have to be written with the autoscheduler in mind.

**Comment by zvookin** [comment]
It was never a question of something happening automatically for legacy code. The high level goal was to raise the level of abstraction available in Halide and to allow writing componentize code. I'd say we still have little data one way on the other how these mechanisms work for that goal.

Re: autoscheduling, the question is what a pipeline written with the autoscheduler in mind should look like. I presume we want to allow manual scheduling and autoscheduling, at least in an entirely one or entirely the other fashion. How do we write a pipeline that does this? It could just be a generator param that turns off applying the manual scheduling, but at the very least we should standardize the name. I think Steven may have already done that.

**Comment by steven-johnson** [comment]
I lumped several issues into one here; there are arguably at least three distinct ones:
- Kill ScheduleParam<> and add GeneratorParam<LoopLevel>
- Remove the separate `schedule()` method.
- Ensure that autoscheduling is adequately accounted for.

Of these three, the first seems non-controversial based on offline and gitter comments, so I'll move forward with PRs on that front; the third is partially implemented already, but needs more thorough usage to verify its seaworthiness in real usage.
